File: apps/api/models/role.py
"""
Role models for dynamic role-based access control (RBAC)
Mirrors Zod schemas from @toko/shared-types/src/schemas/role.ts
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, Dict, Any, Literal
from datetime import datetime
from enum import Enum


# System Role IDs (constants)
class SystemRoleID:
    SUPER_ADMIN = "super_admin"
    OWNER = "owner"
    MEMBER = "member"


# Role Level Constants
class RoleLevel:
    SUPER_ADMIN = 100
    OWNER = 90
    ADMIN = 70
    MANAGER = 50
    STAFF = 30
    VIEWER = 10


# User Permissions Schema (mirrors TypeScript UserPermissions)
class UserPermissions(BaseModel):
    # Products
    canCreateProducts: bool = False
    canEditProducts: bool = False
    canDeleteProducts: bool = False
    canViewProducts: bool = True

    # Orders/Sales
    canCreateOrders: bool = False
    canEditOrders: bool = False
    canDeleteOrders: bool = False
    canViewOrders: bool = True
    canRefundOrders: bool = False

    # Customers
    canManageCustomers: bool = False
    canViewCustomers: bool = True

    # Inventory
    canManageInventory: bool = False
    canViewInventory: bool = True

    # Reports
    canViewReports: bool = False
    canExportReports: bool = False

    # Settings
    canManageSettings: bool = False
    canManageUsers: bool = False
    canManagePayments: bool = False

    class Config:
        # Allow field names to use camelCase (matching TypeScript)
        populate_by_name = True


# System Role Schema (static, read-only)
class SystemRole(BaseModel):
    id: str  # 'super_admin', 'owner', 'member'
    name: str
    description: str
    level: int = Field(ge=90, le=100, description="System roles are level 90+")
    permissions: UserPermissions
    isSystemRole: Literal[True] = True
    isActive: bool = True
    createdAt: datetime
    updatedAt: datetime

    class Config:
        populate_by_name = True


# Tenant Role Schema (dynamic, tenant-specific)
class TenantRole(BaseModel):
    id: str  # UUID
    tenantId: str  # UUID
    name: str = Field(min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)

    # Hierarchy
    level: int = Field(ge=1, le=89, description="Custom roles are level 1-89")
    inheritsFrom: Optional[str] = None  # Parent role ID

    # Permissions
    permissions: UserPermissions

    # Metadata
    isCustom: Literal[True] = True
    isActive: bool = True
    createdBy: str  # User ID
    createdAt: datetime
    updatedAt: datetime

    class Config:
        populate_by_name = True

    @validator('level')
    def validate_level(cls, v):
        if v < 1 or v > 89:
            raise ValueError('Custom role level must be between 1 and 89')
        return v


# Create Tenant Role Input
class CreateTenantRoleInput(BaseModel):
    tenantId: str
    name: str = Field(min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    level: int = Field(ge=1, le=89)
    inheritsFrom: Optional[str] = None
    permissions: UserPermissions

    class Config:
        populate_by_name = True

    @validator('level')
    def validate_level(cls, v):
        if v < 1 or v > 89:
            raise ValueError('Custom role level must be between 1 and 89')
        return v


# Update Tenant Role Input
class UpdateTenantRoleInput(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    level: Optional[int] = Field(None, ge=1, le=89)
    inheritsFrom: Optional[str] = None
    permissions: Optional[Dict[str, bool]] = None  # Partial permissions
    isActive: Optional[bool] = None

    class Config:
        populate_by_name = True

    @validator('level')
    def validate_level(cls, v):
        if v is not None and (v < 1 or v > 89):
            raise ValueError('Custom role level must be between 1 and 89')
        return v


# Role Query Schema
class RoleQuery(BaseModel):
    tenantId: str
    isCustom: Optional[bool] = None
    isActive: Optional[bool] = None
    minLevel: Optional[int] = Field(None, ge=1, le=100)
    maxLevel: Optional[int] = Field(None, ge=1, le=100)
    search: Optional[str] = None  # Search by name or description
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=20, ge=1, le=100)

    class Config:
        populate_by_name = True


# Clone Role Input
class CloneRoleInput(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    tenantId: str

    class Config:
        populate_by_name = True


# Helper function to get default permissions by system role ID
def get_default_permissions_by_system_role(role_id: str) -> UserPermissions:
    """Get default permissions for system roles"""

    all_permissions = UserPermissions(
        canCreateProducts=True,
        canEditProducts=True,
        canDeleteProducts=True,
        canViewProducts=True,
        canCreateOrders=True,
        canEditOrders=True,
        canDeleteOrders=True,
        canViewOrders=True,
        canRefundOrders=True,
        canManageCustomers=True,
        canViewCustomers=True,
        canManageInventory=True,
        canViewInventory=True,
        canViewReports=True,
        canExportReports=True,
        canManageSettings=True,
        canManageUsers=True,
        canManagePayments=True,
    )

    base_permissions = UserPermissions(
        canViewProducts=True,
        canViewOrders=True,
        canViewCustomers=True,
        canViewInventory=True,
    )

    if role_id in [SystemRoleID.SUPER_ADMIN, SystemRoleID.OWNER]:
        return all_permissions
    elif role_id == SystemRoleID.MEMBER:
        return base_permissions
    else:
        return base_permissions


# Helper function to merge permissions
def merge_permissions(parent: UserPermissions, child: Dict[str, bool]) -> UserPermissions:
    """Merge parent permissions with child overrides"""
    parent_dict = parent.dict()
    merged = {**parent_dict, **child}
    return UserPermissions(**merged)


# Helper function to check permission by level
def has_permission_by_level(user_level: int, required_level: int) -> bool:
    """Check if user level meets required level (higher = more authority)"""
    return user_level >= required_level


# Default System Roles (seed data)
DEFAULT_SYSTEM_ROLES = [
    {
        "id": SystemRoleID.SUPER_ADMIN,
        "name": "Super Admin",
        "description": "Platform administrator with full access to all tenants and system settings",
        "level": RoleLevel.SUPER_ADMIN,
        "permissions": get_default_permissions_by_system_role(SystemRoleID.SUPER_ADMIN).dict(),
        "isSystemRole": True,
        "isActive": True,
    },
    {
        "id": SystemRoleID.OWNER,
        "name": "Owner",
        "description": "Tenant owner with full access to all tenant features and settings",
        "level": RoleLevel.OWNER,
        "permissions": get_default_permissions_by_system_role(SystemRoleID.OWNER).dict(),
        "isSystemRole": True,
        "isActive": True,
    },
    {
        "id": SystemRoleID.MEMBER,
        "name": "Member",
        "description": "Basic tenant member with view-only access",
        "level": RoleLevel.VIEWER,
        "permissions": get_default_permissions_by_system_role(SystemRoleID.MEMBER).dict(),
        "isSystemRole": True,
        "isActive": True,
    },
]


# Role Templates (starting points for custom roles)
ROLE_TEMPLATES = {
    "ADMIN": {
        "name": "Admin",
        "description": "Administrator with most permissions except ownership transfer",
        "level": RoleLevel.ADMIN,
        "permissions": UserPermissions(
            canCreateProducts=True,
            canEditProducts=True,
            canDeleteProducts=True,
            canViewProducts=True,
            canCreateOrders=True,
            canEditOrders=True,
            canDeleteOrders=True,
            canViewOrders=True,
            canRefundOrders=True,
            canManageCustomers=True,
            canViewCustomers=True,
            canManageInventory=True,
            canViewInventory=True,
            canViewReports=True,
            canExportReports=True,
            canManageUsers=True,
        ).dict(),
    },
    "MANAGER": {
        "name": "Manager",
        "description": "Store manager with product and order management permissions",
        "level": RoleLevel.MANAGER,
        "permissions": UserPermissions(
            canCreateProducts=True,
            canEditProducts=True,
            canViewProducts=True,
            canCreateOrders=True,
            canEditOrders=True,
            canViewOrders=True,
            canManageCustomers=True,
            canViewCustomers=True,
            canManageInventory=True,
            canViewInventory=True,
            canViewReports=True,
        ).dict(),
    },
    "CASHIER": {
        "name": "Cashier",
        "description": "Cashier with order creation and basic customer management",
        "level": RoleLevel.STAFF,
        "permissions": UserPermissions(
            canViewProducts=True,
            canCreateOrders=True,
            canEditOrders=True,
            canViewOrders=True,
            canManageCustomers=True,
            canViewCustomers=True,
        ).dict(),
    },
    "INVENTORY_MANAGER": {
        "name": "Inventory Manager",
        "description": "Specialized role for inventory and product management",
        "level": RoleLevel.STAFF,
        "permissions": UserPermissions(
            canCreateProducts=True,
            canEditProducts=True,
            canViewProducts=True,
            canManageInventory=True,
            canViewInventory=True,
            canViewReports=True,
        ).dict(),
    },
}

File: packages/shared-types/src/schemas/role.ts
import { z } from 'zod';
import { UserPermissionsSchema } from './user';

/**
 * System Role IDs (static, predefined)
 * These are platform-wide roles that cannot be modified or deleted
 */
export const SYSTEM_ROLE_IDS = {
  SUPER_ADMIN: 'super_admin', // Platform administrator
  OWNER: 'owner', // Tenant owner
  MEMBER: 'member', // Basic tenant member
} as const;

/**
 * Role Level Constants
 * Higher level = more permissions
 * Custom roles can be 1-89, system roles are 90+
 */
export const ROLE_LEVELS = {
  SUPER_ADMIN: 100,
  OWNER: 90,
  ADMIN: 70,
  MANAGER: 50,
  STAFF: 30,
  VIEWER: 10,
} as const;

/**
 * System Role Schema (static, read-only)
 * These roles are predefined and cannot be modified by users
 */
export const SystemRoleSchema = z.object({
  id: z.string(), // 'super_admin', 'owner', 'member'
  name: z.string(),
  description: z.string(),
  level: z.number().int().min(90).max(100), // System roles are level 90+
  permissions: UserPermissionsSchema,
  isSystemRole: z.literal(true),
  isActive: z.boolean().default(true),
  createdAt: z.date(),
  updatedAt: z.date(),
});
export type SystemRole = z.infer<typeof SystemRoleSchema>;

/**
 * Tenant Role Schema (dynamic, tenant-specific)
 * Custom roles created by tenant owners/admins
 */
export const TenantRoleSchema = z.object({
  id: z.string().uuid(),
  tenantId: z.string().uuid(),
  name: z.string().min(1).max(100),
  description: z.string().max(500).optional(),

  // Hierarchy
  level: z.number().int().min(1).max(89), // Custom roles are level 1-89
  inheritsFrom: z.string().optional(), // Parent role ID (system or custom)

  // Permissions (merged with inherited permissions)
  permissions: UserPermissionsSchema,

  // Metadata
  isCustom: z.literal(true),
  isActive: z.boolean().default(true),
  createdBy: z.string(), // User ID who created this role
  createdAt: z.date(),
  updatedAt: z.date(),
});
export type TenantRole = z.infer<typeof TenantRoleSchema>;

/**
 * Combined Role Schema (union of system and tenant roles)
 * Used when querying roles regardless of type
 */
export const RoleSchema = z.union([SystemRoleSchema, TenantRoleSchema]);
export type Role = z.infer<typeof RoleSchema>;

/**
 * Create Tenant Role Input
 */
export const CreateTenantRoleInputSchema = z.object({
  name: z.string().min(1).max(100),
  description: z.string().max(500).optional(),
  level: z.number().int().min(1).max(89),
  inheritsFrom: z.string().optional(),
  permissions: UserPermissionsSchema,
});
export type CreateTenantRoleInput = z.infer<typeof CreateTenantRoleInputSchema>;

/**
 * Update Tenant Role Input
 */
export const UpdateTenantRoleInputSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).optional(),
  level: z.number().int().min(1).max(89).optional(),
  inheritsFrom: z.string().optional(),
  permissions: UserPermissionsSchema.partial().optional(),
  isActive: z.boolean().optional(),
});
export type UpdateTenantRoleInput = z.infer<typeof UpdateTenantRoleInputSchema>;

/**
 * Role Query Schema (for filtering/searching roles)
 */
export const RoleQuerySchema = z.object({
  tenantId: z.string().uuid().optional(),
  isActive: z.boolean().optional(),
  isCustom: z.boolean().optional(),
  minLevel: z.number().int().min(1).max(100).optional(),
  maxLevel: z.number().int().min(1).max(100).optional(),
  search: z.string().optional(), // Search by name or description
});
export type RoleQuery = z.infer<typeof RoleQuerySchema>;

/**
 * Helper function to get default permissions by system role ID
 */
export function getDefaultPermissionsBySystemRole(roleId: string): typeof UserPermissionsSchema._type {
  const allPermissions = {
    canCreateProducts: true,
    canEditProducts: true,
    canDeleteProducts: true,
    canViewProducts: true,
    canCreateOrders: true,
    canEditOrders: true,
    canDeleteOrders: true,
    canViewOrders: true,
    canRefundOrders: true,
    canManageCustomers: true,
    canViewCustomers: true,
    canManageInventory: true,
    canViewInventory: true,
    canViewReports: true,
    canExportReports: true,
    canManageSettings: true,
    canManageUsers: true,
    canManagePayments: true,
  };

  const basePermissions = {
    canCreateProducts: false,
    canEditProducts: false,
    canDeleteProducts: false,
    canViewProducts: true,
    canCreateOrders: false,
    canEditOrders: false,
    canDeleteOrders: false,
    canViewOrders: true,
    canRefundOrders: false,
    canManageCustomers: false,
    canViewCustomers: true,
    canManageInventory: false,
    canViewInventory: true,
    canViewReports: false,
    canExportReports: false,
    canManageSettings: false,
    canManageUsers: false,
    canManagePayments: false,
  };

  switch (roleId) {
    case SYSTEM_ROLE_IDS.SUPER_ADMIN:
    case SYSTEM_ROLE_IDS.OWNER:
      return allPermissions;
    case SYSTEM_ROLE_IDS.MEMBER:
      return basePermissions;
    default:
      return basePermissions;
  }
}

/**
 * Helper function to merge permissions from parent role
 * Child role inherits all parent permissions and can add more
 */
export function mergePermissions(
  parentPermissions: typeof UserPermissionsSchema._type,
  childPermissions: Partial<typeof UserPermissionsSchema._type>
): typeof UserPermissionsSchema._type {
  return {
    ...parentPermissions,
    ...childPermissions,
  };
}

/**
 * Helper function to check if user has permission based on role level
 * Higher level roles automatically have permissions of lower level roles
 */
export function hasPermissionByLevel(userLevel: number, requiredLevel: number): boolean {
  return userLevel >= requiredLevel;
}

/**
 * System Role Definitions (seed data)
 * These should be created when initializing the database
 */
export const DEFAULT_SYSTEM_ROLES: Omit<SystemRole, 'createdAt' | 'updatedAt'>[] = [
  {
    id: SYSTEM_ROLE_IDS.SUPER_ADMIN,
    name: 'Super Admin',
    description: 'Platform administrator with full access to all tenants and system settings',
    level: ROLE_LEVELS.SUPER_ADMIN,
    permissions: getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.SUPER_ADMIN),
    isSystemRole: true,
    isActive: true,
  },
  {
    id: SYSTEM_ROLE_IDS.OWNER,
    name: 'Owner',
    description: 'Tenant owner with full access to all tenant features and settings',
    level: ROLE_LEVELS.OWNER,
    permissions: getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.OWNER),
    isSystemRole: true,
    isActive: true,
  },
  {
    id: SYSTEM_ROLE_IDS.MEMBER,
    name: 'Member',
    description: 'Basic tenant member with view-only access',
    level: ROLE_LEVELS.VIEWER,
    permissions: getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.MEMBER),
    isSystemRole: true,
    isActive: true,
  },
];

/**
 * Example Custom Role Templates
 * These can be used as starting points for custom roles
 */
export const ROLE_TEMPLATES = {
  ADMIN: {
    name: 'Admin',
    description: 'Administrator with most permissions except ownership transfer',
    level: ROLE_LEVELS.ADMIN,
    permissions: {
      ...getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.MEMBER),
      canCreateProducts: true,
      canEditProducts: true,
      canDeleteProducts: true,
      canCreateOrders: true,
      canEditOrders: true,
      canDeleteOrders: true,
      canRefundOrders: true,
      canManageCustomers: true,
      canManageInventory: true,
      canViewReports: true,
      canExportReports: true,
      canManageUsers: true,
    },
  },
  MANAGER: {
    name: 'Manager',
    description: 'Store manager with product and order management permissions',
    level: ROLE_LEVELS.MANAGER,
    permissions: {
      ...getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.MEMBER),
      canCreateProducts: true,
      canEditProducts: true,
      canCreateOrders: true,
      canEditOrders: true,
      canManageCustomers: true,
      canManageInventory: true,
      canViewReports: true,
    },
  },
  CASHIER: {
    name: 'Cashier',
    description: 'Cashier with order creation and basic customer management',
    level: ROLE_LEVELS.STAFF,
    permissions: {
      ...getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.MEMBER),
      canCreateOrders: true,
      canEditOrders: true,
      canManageCustomers: true,
    },
  },
  INVENTORY_MANAGER: {
    name: 'Inventory Manager',
    description: 'Specialized role for inventory and product management',
    level: ROLE_LEVELS.STAFF,
    permissions: {
      ...getDefaultPermissionsBySystemRole(SYSTEM_ROLE_IDS.MEMBER),
      canCreateProducts: true,
      canEditProducts: true,
      canManageInventory: true,
      canViewReports: true,
    },
  },
} as const;
